File: examples/go/cases/tasks/task_list/task_list.domain.case.go
package task_list

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"app-protocol/examples/go/core"
)

type TaskStatus string

const (
	TaskStatusTodo  TaskStatus = "todo"
	TaskStatusDoing TaskStatus = "doing"
	TaskStatusDone  TaskStatus = "done"
)

type Task struct {
	ID          string     `json:"id"`
	Title       string     `json:"title"`
	Description string     `json:"description,omitempty"`
	Status      TaskStatus `json:"status"`
	CreatedAt   string     `json:"createdAt"`
	UpdatedAt   string     `json:"updatedAt"`
}

type TaskListInput struct{}

type TaskListOutput struct {
	Tasks []Task `json:"tasks"`
}

type Domain struct {
	core.DomainCaseSupport
}

func NewDomain() *Domain {
	return &Domain{}
}

func (d *Domain) CaseName() string {
	return "task_list"
}

func (d *Domain) Description() string {
	return "Lists persisted task cards for board rendering."
}

func (d *Domain) InputSchema() core.AppSchema {
	return core.AppSchema{
		"type":                 "object",
		"properties":           map[string]any{},
		"additionalProperties": false,
	}
}

func (d *Domain) OutputSchema() core.AppSchema {
	return core.AppSchema{
		"type": "object",
		"properties": map[string]any{
			"tasks": map[string]any{
				"type": "array",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"id":          map[string]any{"type": "string"},
						"title":       map[string]any{"type": "string"},
						"description": map[string]any{"type": "string"},
						"status": map[string]any{
							"type": "string",
							"enum": []string{"todo", "doing", "done"},
						},
						"createdAt": map[string]any{"type": "string"},
						"updatedAt": map[string]any{"type": "string"},
					},
					"required":             []string{"id", "title", "status", "createdAt", "updatedAt"},
					"additionalProperties": false,
				},
			},
		},
		"required":             []string{"tasks"},
		"additionalProperties": false,
	}
}

func (d *Domain) ParseInput(value any) (TaskListInput, error) {
	if value == nil {
		return TaskListInput{}, nil
	}

	record, ok := value.(map[string]any)
	if !ok {
		encoded, err := json.Marshal(value)
		if err != nil {
			return TaskListInput{}, fmt.Errorf("input must be an object")
		}

		if err := json.Unmarshal(encoded, &record); err != nil {
			return TaskListInput{}, fmt.Errorf("input must be an object")
		}
	}

	if len(record) > 0 {
		return TaskListInput{}, fmt.Errorf("task_list does not accept filters in v1")
	}

	return TaskListInput{}, nil
}

func (d *Domain) Validate(_ TaskListInput) error {
	return nil
}

func (d *Domain) ValidateOutput(output TaskListOutput) error {
	for index, task := range output.Tasks {
		if err := d.ValidateTask(task); err != nil {
			return fmt.Errorf("task_list.output.tasks[%d]: %w", index, err)
		}
	}

	return nil
}

func (d *Domain) ValidateTask(task Task) error {
	if strings.TrimSpace(task.ID) == "" {
		return fmt.Errorf("id must be a non-empty string")
	}

	if strings.TrimSpace(task.Title) == "" {
		return fmt.Errorf("title must be a non-empty string")
	}

	switch task.Status {
	case TaskStatusTodo, TaskStatusDoing, TaskStatusDone:
	default:
		return fmt.Errorf("status must be one of todo, doing, done")
	}

	if strings.TrimSpace(task.CreatedAt) == "" {
		return fmt.Errorf("createdAt must be a non-empty string")
	}

	if strings.TrimSpace(task.UpdatedAt) == "" {
		return fmt.Errorf("updatedAt must be a non-empty string")
	}

	return nil
}

func (d *Domain) Invariants() []string {
	return []string{
		"Only todo, doing, and done are valid task statuses.",
		"Listing tasks never mutates the persisted store.",
		"The response order is deterministic for the same persisted dataset.",
	}
}

func (d *Domain) Examples() []core.DomainExample {
	return []core.DomainExample{
		{
			Name:        "empty_board",
			Description: "No persisted tasks yet.",
			Input:       TaskListInput{},
			Output: TaskListOutput{
				Tasks: []Task{},
			},
		},
		{
			Name:        "board_with_cards",
			Description: "Returns tasks already persisted in the board.",
			Input:       TaskListInput{},
			Output: TaskListOutput{
				Tasks: []Task{
					{
						ID:        "task_002",
						Title:     "Prepare release notes",
						Status:    TaskStatusTodo,
						CreatedAt: "2026-03-18T12:10:00.000Z",
						UpdatedAt: "2026-03-18T12:10:00.000Z",
					},
					{
						ID:          "task_001",
						Title:       "Ship the Go example",
						Description: "Wire the APP cases in the portal.",
						Status:      TaskStatusDoing,
						CreatedAt:   "2026-03-18T12:00:00.000Z",
						UpdatedAt:   "2026-03-18T12:30:00.000Z",
					},
				},
			},
		},
	}
}

func (d *Domain) Definition() core.DomainDefinition {
	return core.DomainDefinition{
		CaseName:     d.CaseName(),
		Description:  d.Description(),
		InputSchema:  d.InputSchema(),
		OutputSchema: d.OutputSchema(),
		Invariants:   d.Invariants(),
		ValueObjects: d.ValueObjects(),
		Enums:        d.Enums(),
		Examples:     d.Examples(),
	}
}

func (d *Domain) SortTasks(tasks []Task) []Task {
	sorted := append([]Task(nil), tasks...)
	sort.SliceStable(sorted, func(left int, right int) bool {
		return sorted[left].CreatedAt > sorted[right].CreatedAt
	})
	return sorted
}

func (d *Domain) Test() error {
	if _, err := d.ParseInput(map[string]any{"status": "todo"}); err == nil {
		return fmt.Errorf("test: task_list input must reject filters")
	}

	example, ok := d.Examples()[1].Output.(TaskListOutput)
	if !ok {
		return fmt.Errorf("test: board_with_cards example must exist")
	}

	if err := d.ValidateOutput(example); err != nil {
		return err
	}

	if err := d.ValidateOutput(TaskListOutput{
		Tasks: []Task{
			{
				ID:        "bad",
				Title:     "Invalid task",
				Status:    "invalid",
				CreatedAt: "2026-03-18T12:00:00.000Z",
				UpdatedAt: "2026-03-18T12:00:00.000Z",
			},
		},
	}); err == nil {
		return fmt.Errorf("test: validateOutput must reject invalid task status")
	}

	return nil
}

File: examples/kotlin/cases/tasks/task_list/task_list.domain.case.kt
package app.protocol.examples.kotlin.cases.tasks.task_list

import app.protocol.examples.kotlin.core.AppSchema
import app.protocol.examples.kotlin.core.BaseDomainCase
import app.protocol.examples.kotlin.core.DomainExample
import kotlinx.serialization.Serializable

private val taskStatusValues = listOf("todo", "doing", "done")

@Serializable
data class Task(
    val id: String,
    val title: String,
    val description: String? = null,
    val status: String,
    val createdAt: String,
    val updatedAt: String,
)

@Serializable
data object TaskListInput

@Serializable
data class TaskListOutput(
    val tasks: List<Task>,
)

fun assertTaskCollection(value: List<Task>, source: String) {
    value.forEachIndexed { index, task ->
        require(task.id.isNotBlank()) { "$source[$index].id must be a non-empty string" }
        require(task.title.isNotBlank()) { "$source[$index].title must be a non-empty string" }
        require(task.createdAt.isNotBlank()) { "$source[$index].createdAt must be a non-empty string" }
        require(task.updatedAt.isNotBlank()) { "$source[$index].updatedAt must be a non-empty string" }
        require(task.status in taskStatusValues) {
            "$source[$index].status must be one of ${taskStatusValues.joinToString()}"
        }
    }
}

class TaskListDomain : BaseDomainCase<TaskListInput, TaskListOutput>() {
    override fun caseName(): String = "task_list"

    override fun description(): String = "Lists persisted task cards for board rendering."

    override fun inputSchema(): AppSchema =
        AppSchema(
            type = "object",
            properties = emptyMap(),
            additionalProperties = false,
        )

    override fun outputSchema(): AppSchema =
        AppSchema(
            type = "object",
            properties = mapOf(
                "tasks" to AppSchema(
                    type = "array",
                    items = AppSchema(
                        type = "object",
                        properties = mapOf(
                            "id" to AppSchema(type = "string"),
                            "title" to AppSchema(type = "string"),
                            "description" to AppSchema(type = "string"),
                            "status" to AppSchema(type = "string", enum = taskStatusValues),
                            "createdAt" to AppSchema(type = "string"),
                            "updatedAt" to AppSchema(type = "string"),
                        ),
                        required = listOf("id", "title", "status", "createdAt", "updatedAt"),
                        additionalProperties = false,
                    ),
                ),
            ),
            required = listOf("tasks"),
            additionalProperties = false,
        )

    override fun validate(input: TaskListInput) {
    }

    fun validateOutput(output: TaskListOutput) {
        assertTaskCollection(output.tasks, "task_list.output.tasks")
    }

    override fun invariants(): List<String> =
        listOf(
            "Only todo, doing, done are valid task statuses.",
            "Listing tasks never mutates the persisted store.",
            "The response order is deterministic for the same persisted dataset.",
        )

    override fun examples(): List<DomainExample<TaskListInput, TaskListOutput>> =
        listOf(
            DomainExample(
                name = "empty_board",
                description = "No persisted tasks yet.",
                input = TaskListInput,
                output = TaskListOutput(tasks = emptyList()),
            ),
            DomainExample(
                name = "board_with_cards",
                description = "Returns tasks already persisted in the board.",
                input = TaskListInput,
                output = TaskListOutput(
                    tasks = listOf(
                        Task(
                            id = "task_002",
                            title = "Prepare release notes",
                            status = "todo",
                            createdAt = "2026-03-18T12:10:00Z",
                            updatedAt = "2026-03-18T12:10:00Z",
                        ),
                        Task(
                            id = "task_001",
                            title = "Ship the Kotlin example",
                            description = "Wire the first APP cases in the portal.",
                            status = "doing",
                            createdAt = "2026-03-18T12:00:00Z",
                            updatedAt = "2026-03-18T12:30:00Z",
                        ),
                    ),
                ),
            ),
        )

    override suspend fun test() {
        validate(TaskListInput)

        val example =
            examples().firstOrNull { it.name == "board_with_cards" }
                ?: error("test: board_with_cards example must exist")
        validateOutput(example.output ?: error("test: board_with_cards output must exist"))

        val invalidStatus =
            runCatching {
                validateOutput(
                    TaskListOutput(
                        tasks = listOf(
                            Task(
                                id = "bad",
                                title = "Invalid task",
                                status = "invalid",
                                createdAt = "2026-03-18T12:00:00Z",
                                updatedAt = "2026-03-18T12:00:00Z",
                            ),
                        ),
                    ),
                )
            }
        check(invalidStatus.isFailure) {
            "test: validateOutput must reject invalid task status"
        }
    }
}
